gocourse03/mouseMetropolis/main.go

```diff
@@ -0,0 +1,197 @@
+/*
+У зоопарку є окрема кімната «малі гризуни», де знаходиться кілька сімейств кількох видів гризунів. У кожного є чіп з
+його ID і видом. Для них у цій кімнаті побудована мережа лабіринтів. У цьому лабіринті є один датчик, через який
+впродовж дня проходять сотні гризунів, а до ночі вони розходяться по своїх куточках. Для працівників зоопарку треба
+зробити програму, яка буде давати звіт, де знаходився кожний гризун на початку дня і де зупинився
+наприкінці, а також зберігати історію його рухів через датчик. У програмі може бути тип/структура для певного виду
+гризуна, набори (слайси) цих гризунів, і слайс для зберігання руху гризунів. Також стан сектору лабіринту: скільки
+і які гризуни знаходиться там у певний момент. Треба симулювати рух гризунів через головний датчик і датчики секторів.
+Виконати операції видалення, пошуку і додавання в слайд.
+*/
+package main
+
+import (
+	"fmt"
+	"math/rand/v2"
+	"time"
+
+	"github.com/brianvoe/gofakeit/v7"
+)
+
+type (
+	Sector     string
+	RodentType string
+	FromTo     [2]Sector
+
+	Movement struct {
+		Time time.Time
+		FromTo
+	}
+
+	MiceRoom struct {
+		Rodents []Rodent
+	}
+
+	Rodent struct {
+		ID      int
+		Type    RodentType
+		History []Movement
+	}
+)
+
+const (
+	rodentCount = 15
+
+	rat   RodentType = "rat"
+	mouse RodentType = "mouse"
+
+	sectorA Sector = "sectorA"
+	sectorB Sector = "sectorB"
+	sectorC Sector = "sectorC"
+	sectorD Sector = "sectorD"
+	sectorE Sector = "sectorE"
+)
+
+// this constructor randomly fills the typeRodent
+func newRodent(id int) *Rodent {
+	typeRodent := mouse
+
+	if rand.N(2) == 1 {
+		typeRodent = rat
+	}
+	// make +1 because the ID cannot be 0
+	return &Rodent{ID: id + 1, Type: typeRodent, History: nil}
+}
+
+func (r *Rodent) addMovement(movement Movement) {
+	r.History = append(r.History, movement)
+}
+
+func (m *MiceRoom) delMovement(t time.Time) bool {
+	delStatus := false
+	for _, rodent := range m.Rodents {
+		for i, m := range rodent.History {
+			if m.Time == t {
+				rodent.History = append(rodent.History[:i], rodent.History[i+1:]...)
+				delStatus = true
+			}
+		}
+	}
+	return delStatus
+}
+
+func (m *MiceRoom) findHistory(id int) map[time.Time][]FromTo {
+	history := make(map[time.Time][]FromTo)
+	for _, rodent := range m.Rodents {
+		if rodent.ID == id {
+			for _, h := range rodent.History {
+				history[h.Time] = append(history[h.Time], h.FromTo)
+			}
+		}
+	}
+	return history
+}
+
+func (r *Rodent) moveRodent(count int) {
+	var from, to Sector
+	if count == 0 {
+		count++
+	}
+
+	for i := 0; i < count; i++ {
+		if r.History == nil {
+			from = sectorA
+		} else {
+			from = r.History[len(r.History)-1].FromTo[1]
+		}
```

Comment:
Так працюватиме?
```suggestion
		from = sectorA
		if r.History != nil {
			from = r.History[len(r.History)-1].FromTo[1]
		}
```

Трішки простіше.